### PR TITLE
ctrld: init at 1.3.3

### PR DESCRIPTION
**Pull request by arthsmn**
## Description of changes

Homepage: https://github.com/Control-D-Inc/ctrld/

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/networking/ctrld.nix [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            name = "Control D - No Ads";
+            timeout = 3000;
+            type = "doq";
+          };
+        };
+      };
+
+      description = ''
+        Configuration for ctrld, see
+        <link xlink:href="https://github.com/Control-D-Inc/ctrld/blob/release-branch-v${pkgs.ctrld.version}/docs/config.md">
```
Comment (pbsds) [inline nixos/modules/services/networking/ctrld.nix:55]: It's commonmark now

```suggestion
        <https://github.com/Control-D-Inc/ctrld/blob/release-branch-v${pkgs.ctrld.version}/docs/config.md>
```

File: nixos/modules/services/networking/ctrld.nix [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        ExecStart = "${getExe pkgs.ctrld} start --config ${cfg.configFile}";
+      };
+      after = [ "network.target" ];
+      before = [ "nss-lookup.target" ];
+      wants = [ "nss-lookup.target" ];
+      wantedBy = [ "multi-user.target" ];
+    };
+  };
+
+  meta.buildDocsInSandbox = false;
```
Comment (pbsds) [inline nixos/modules/services/networking/ctrld.nix:77]: Why is this needed?

**Review by pbsds — COMMENTED** [review]
Could you add a nixos test?

File: nixos/modules/services/networking/ctrld.nix
Comment (pbsds) [inline nixos/modules/services/networking/ctrld.nix:1]: Please set `meta.maintainers`

File: nixos/modules/services/networking/ctrld.nix (hunk 1 above)
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix:55]: Done.

File: nixos/modules/services/networking/ctrld.nix (hunk 2 above)
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix:77]: Because I'm using pkgs.ctrld.version in meta.description.

File: nixos/modules/services/networking/ctrld.nix
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix:1]: Done.

**Comment by arthsmn** [comment]
> Could you add a nixos test?

How so? Do you want to test if the binary can run or the service works?

File: nixos/modules/services/networking/ctrld.nix [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+          "upstream.1" = {
+            bootstrap_ip = "76.76.2.11";
+            endpoint = "p2.freedns.controld.com";
+            name = "Control D - No Ads";
+            timeout = 3000;
+            type = "doq";
+          };
+        };
+      };
```
Comment (pbsds) [inline nixos/modules/services/networking/ctrld.nix:51]: These upstreams strike me as odd defaults for nixos to promote. Could this config perhaps be changed into the example? The commented link to the upstream default_config could be moved into the `description` for the user to view and reference themselves.

File: nixos/modules/services/networking/ctrld.nix [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+with lib;
+
+let
+  cfg = config.services.ctrld;
+  settingsFormat = pkgs.formats.toml {};
+in {
+
+  options.services.ctrld = {
+    enable = mkEnableOption "ctrld service";
+
```
Comment (pbsds) [inline nixos/modules/services/networking/ctrld.nix:12]:
```suggestion

    package = mkPackageOption pkgs "ctrld" { };

```

File: nixos/modules/services/networking/ctrld.nix [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  config = mkIf cfg.enable {
+    environment.etc."ctrld/ctrld.toml".source = settingsFormat.generate "ctrld.toml" cfg.settings;
+
+    systemd.services.ctrld = {
+      description = "ctrld proxy client";
+      startLimitIntervalSec = 5;
+      startLimitBurst = 10;
+      serviceConfig = {
+        ExecStart = "${getExe pkgs.ctrld} start --config ${cfg.configFile}";
```
Comment (pbsds) [inline nixos/modules/services/networking/ctrld.nix:68]:
```suggestion
        ExecStart = "${getExe cfg.package} start --config ${cfg.configFile}";
```

File: nixos/modules/services/networking/ctrld.nix (hunk 3 above)
Comment (pbsds) [reply nixos/modules/services/networking/ctrld.nix:51]: The simplest config imaginable should be the default

**Review by pbsds — COMMENTED** [review]
> How so? Do you want to test if the binary can run or the service works?

Yes, a simple "does the service start and listen on the proper port" is a decent test. You don't have to perform a full test of all possible functionality, that job is for the upstream to ensure is working.

File: nixos/modules/services/networking/ctrld.nix (hunk 4 above)
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix:12]: Done.

File: nixos/modules/services/networking/ctrld.nix (hunk 5 above)
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix]: Done.

File: nixos/modules/services/networking/ctrld.nix (hunk 3 above)
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix:53]: I think an empty config would be ideal in this case. I saw that NextDNS uses an empty config so I think it should be like this.

File: nixos/modules/services/networking/ctrld.nix [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    networking.nameservers = mkDefault [ "127.0.0.1" ];
+
+    environment.etc."ctrld/ctrld.toml".source = settingsFormat.generate "ctrld.toml" cfg.settings;
+
+    systemd.services.ctrld = {
+      description = "ctrld proxy client";
+      startLimitIntervalSec = 5;
+      startLimitBurst = 10;
+      serviceConfig = {
+        ExecStart = "${getExe cfg.package} start --config /etc/ctrld/ctrld.toml";
```
Comment (h7x4) [inline nixos/modules/services/networking/ctrld.nix:74]: Maybe this could be moved straight into the `ExecStart` command?

```suggestion
    systemd.services.ctrld = {
      description = "ctrld proxy client";
      startLimitIntervalSec = 5;
      startLimitBurst = 10;
      serviceConfig = {
        ExecStart = "${getExe cfg.package} start --config ${settingsFormat.generate "ctrld.toml" cfg.settings};
```

File: nixos/modules/services/networking/ctrld.nix [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+with lib;
+
+let
+  cfg = config.services.ctrld;
+  settingsFormat = pkgs.formats.toml {};
+in {
+
+  options.services.ctrld = {
+    enable = mkEnableOption "ctrld service";
```
Comment (h7x4) [inline nixos/modules/services/networking/ctrld.nix:11]: Maybe this could be a bit more descriptive? I think the grammar of the option sounds a bit weird as well, considering the content gets appended to "Whether to enable"

How about something like this?

```suggestion
    enable = mkEnableOption "ctrld, a highly configurable DNS forwarding proxy";
```

File: nixos/modules/services/networking/ctrld.nix [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+      description = ''
+        Configuration for ctrld, see
+        <https://github.com/Control-D-Inc/ctrld/blob/release-branch-v${pkgs.ctrld.version}/docs/config.md>
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    networking.nameservers = mkDefault [ "127.0.0.1" ];
```
Comment (h7x4) [inline nixos/modules/services/networking/ctrld.nix:65]: I'm not too sure we should be as careful as to use `mkDefault` here. I imagine a setting where the user enables this service, forgetting that they set their name server to `8.8.8.8` or whatever, and now everything doesn't work for some reason.

File: nixos/tests/ctrld.nix [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  name = "ctrld";
+
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ arthsmn ];
+  };
+
+  nodes = {
+    # machine = { pkgs, ... }: {
+    #   environment.systemPackages = [ pkgs.dig ];
+    # };
```
Comment (h7x4) [inline nixos/tests/ctrld.nix:12]: Were there any plans for this, or should it be removed?

File: nixos/modules/services/networking/ctrld.nix [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            name = "Control D - No Ads";
+            timeout = 3000;
+            type = "doq";
+          };
+        };
+      };
+
+      description = ''
+        Configuration for ctrld, see
+        <https://github.com/Control-D-Inc/ctrld/blob/release-branch-v${pkgs.ctrld.version}/docs/config.md>
```
[the remark this answers is not in the document]
Comment (h7x4) [reply nixos/modules/services/networking/ctrld.nix:59]: nvm, found the culprit :)

I don't think the benefits of specifying the version surpass those of the docs sandbox. We can probably trust the user to find the correct version of the config themselves, as is being assumed in so many other modules.

[the remark this answers is not in the document]
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix]: Done.

File: nixos/modules/services/networking/ctrld.nix [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    };
+  };
+
+  config = mkIf cfg.enable {
+    networking.nameservers = mkDefault [ "127.0.0.1" ];
+
+    environment.etc."ctrld/ctrld.toml".source = settingsFormat.generate "ctrld.toml" cfg.settings;
+
+    systemd.services.ctrld = {
+      description = "ctrld proxy client";
```
[the remark this answers is not in the document]
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix]: Done.

File: nixos/modules/services/networking/ctrld.nix (hunk 6 above)
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix]: Done.

File: nixos/modules/services/networking/ctrld.nix (hunk 7 above)
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix:11]: Done.

File: nixos/modules/services/networking/ctrld.nix (hunk 8 above)
Comment (arthsmn) [reply nixos/modules/services/networking/ctrld.nix:65]: Ok. Done.

File: nixos/tests/ctrld.nix (hunk 9 above)
Comment (arthsmn) [reply nixos/tests/ctrld.nix]: Sorry, I was testing and forgot to uncomment. Done.

File: pkgs/by-name/ct/ctrld/package.nix [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  ldflags = [
+    "-s"
+    "-w"
+    "-X=main.version=${version}"
+    "-X=main.commit=${src.rev}"
+  ];
+
+  # Tests need internet
+  doCheck = false;
+
```
Comment (pbsds) [inline pkgs/by-name/ct/ctrld/package.nix:28]:
```suggestion

  passthru.tests = nixosTests.ctrld;

```

File: pkgs/by-name/ct/ctrld/package.nix [hunk 13]
```diff
@@ -0,0 +1,36 @@
+{ lib
+, buildGoModule
+, fetchFromGitHub
```
Comment (pbsds) [inline pkgs/by-name/ct/ctrld/package.nix:3]:
```suggestion
, fetchFromGitHub
, nixosTests
```

File: nixos/tests/ctrld.nix [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+  name = "ctrld";
+
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ arthsmn ];
+  };
+
+  nodes = {
+    machine = { pkgs, ... }: {
+      environment.systemPackages = [ pkgs.dig ];
+    };
```
Comment (pbsds) [inline nixos/tests/ctrld.nix:12]: unused?

Either test something like `machine.wait_until_succeeds("dig verify.controld.com @client +short")` or

```suggestion
```

File: nixos/tests/ctrld.nix [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+    client = { ... }: {
```
Comment (pbsds) [inline nixos/tests/ctrld.nix:14]:
```suggestion
    client = { pkgs, ... }: {
```

File: pkgs/by-name/ct/ctrld/package.nix (hunk 12 above)
Comment (arthsmn) [reply pkgs/by-name/ct/ctrld/package.nix:28]: Done.

File: pkgs/by-name/ct/ctrld/package.nix (hunk 13 above)
Comment (arthsmn) [reply pkgs/by-name/ct/ctrld/package.nix:3]: Done.

File: nixos/tests/ctrld.nix (hunk 14 above)
Comment (arthsmn) [reply nixos/tests/ctrld.nix:12]: Like this?

File: nixos/tests/ctrld.nix (hunk 15 above)
Comment (arthsmn) [reply nixos/tests/ctrld.nix:14]: Done.

File: pkgs/by-name/ct/ctrld/package.nix [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+  vendorHash = "sha256-il3jA4Dtfs1GMyL3IEgmzRDOIsKViovBqhtsvzwgafQ=";
+
+  ldflags = [
+    "-s"
+    "-w"
+    "-X=main.version=${version}"
+    "-X=main.commit=${src.rev}"
+  ];
+
```
Comment (pbsds) [inline pkgs/by-name/ct/ctrld/package.nix:26]: Seems like this got removed

```suggestion

  # requires network
  doCheck = false;

```

Comment (arthsmn) [reply pkgs/by-name/ct/ctrld/package.nix:26]: Done.

File: nixos/tests/ctrld.nix [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+    client = { pkgs, ... }: {
+      environment.systemPackages = [ pkgs.dig ];
+      services.ctrld.enable = true;
+    };
+  };
+
+  # https://github.com/Control-D-Inc/ctrld#basic-run-mode
+  testScript = ''
+    client.wait_for_unit("ctrld")
+    client.wait_until_succeeds("dig verify.controld.com @client +short")
```
Comment (pbsds) [inline nixos/tests/ctrld.nix:19]: from ofborg:
```
> Exception: action timed out after 900 seconds
```

likely fix:

```suggestion
    client.wait_until_succeeds("dig verify.controld.com @127.0.0.1 +short")
```

For `client` to work, ctrld may need to be listening on `0.0.0.0`

File: pkgs/by-name/ct/ctrld/package.nix [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+  # Requires network
+    doCheck = false;
```
Comment (pbsds) [inline pkgs/by-name/ct/ctrld/package.nix:28]:
```suggestion
  doCheck = false;
```

File: nixos/tests/ctrld.nix (hunk 17 above)
Comment (arthsmn) [reply nixos/tests/ctrld.nix]: Done.

File: pkgs/by-name/ct/ctrld/package.nix (hunk 18 above)
Comment (arthsmn) [reply pkgs/by-name/ct/ctrld/package.nix]: Done.

**Comment by pbsds** [comment]
The tests are failing, both on ofborg and locally. I took it for a spin locally but didn't see anything obvious. Does the test perhaps need something in `settings`?

**Comment by arthsmn** [comment]
> The tests are failing, both on ofborg and locally. I took it for a spin locally but didn't see anything obvious. Does the test perhaps need something in `settings`?

Sorry! Forgot to add the default settings.

**Comment by pbsds** [comment]
Still getting the same error

**Comment by arthsmn** [comment]
I no longer have time/interest in adding this package, so I'll close this PR.
